test/OAuth2ManagerTests/OAuth2APITests.cpp

```diff
@@ -0,0 +1,1225 @@
+using namespace winrt::Windows::Security::Cryptography;
+using namespace winrt::Windows::Security::Cryptography::Core;
+using namespace winrt::Windows::Storage::Streams;
+
+namespace TB = ::Test::Bootstrap;
+namespace TP = ::Test::Packages;
+EXTERN_C IMAGE_DOS_HEADER __ImageBase;
+
+namespace OAuth2ManagerTests
+{
+    class OAuth2APITests
+    {
+    public:
+        BEGIN_TEST_CLASS(OAuth2APITests)
+            TEST_CLASS_PROPERTY(L"ThreadingModel", L"MTA")
+            END_TEST_CLASS()
+
+        TEST_METHOD_SETUP(MethodInit)
+        {
+            VERIFY_IS_TRUE(TP::IsPackageRegistered_WindowsAppRuntimeFramework());
+            return true;
+        }
+
+        TEST_METHOD_CLEANUP(MethodUninit)
+        {
+            VERIFY_IS_TRUE(TP::IsPackageRegistered_WindowsAppRuntimeFramework());
+            return true;
+        }
+
+        TEST_CLASS_SETUP(Setup)
+        {
+            // Initialize the Windows Runtime
+            winrt::init_apartment();
+            Test::Bootstrap::Setup();
+            Test::Packages::WapProj::AddPackage(Test::TAEF::GetDeploymentDir(), L"OAuthTestAppPackage", L".msix");
+
+            // Detour ShellExecuteW
+            ::DetourTransactionBegin();
+            ::DetourUpdateThread(::GetCurrentThread());
+            if (auto err = ::DetourAttach(reinterpret_cast<void**>(&RealShellExecuteW), &DetouredShellExecuteW))
+            {
+                Log::Error(WEX::Common::String().Format(L"DetourAttach failed: %d", err));
+                ::DetourTransactionAbort();
+                return false;
+            }
+            ::DetourTransactionCommit();
+
+            // Initialize the HTTP server we use for token requests
+            VERIFY_WIN32_SUCCEEDED(::HttpInitialize(HTTPAPI_VERSION_2, HTTP_INITIALIZE_SERVER, nullptr));
+            VERIFY_WIN32_SUCCEEDED(::HttpCreateRequestQueue(HTTPAPI_VERSION_2, nullptr, nullptr, 0, &m_requestQueue));
+            VERIFY_WIN32_SUCCEEDED(::HttpCreateServerSession(HTTPAPI_VERSION_2, &m_serverSessionId, 0));
+            VERIFY_WIN32_SUCCEEDED(::HttpCreateUrlGroup(m_serverSessionId, &m_urlGroup, 0));
+
+            HTTP_BINDING_INFO bindingInfo = {};
+            bindingInfo.Flags.Present = 1;
+            bindingInfo.RequestQueueHandle = m_requestQueue;
+            VERIFY_WIN32_SUCCEEDED(::HttpSetUrlGroupProperty(m_urlGroup, HttpServerBindingProperty, &bindingInfo,
+                static_cast<ULONG>(sizeof(bindingInfo))));
+
+            // Find an open port; note that ports in the low 50000s are frequently claimed, hence the large iteration bounds
+            ULONG err = 0;
+            for (std::uint16_t i = 0; i < 500; ++i)
+            {
+                wchar_t buffer[18 + 5 + 1];
+                std::swprintf(buffer, std::size(buffer), L"http://127.0.0.1:%d/", m_serverPort);
+
+                err = ::HttpAddUrlToUrlGroup(m_urlGroup, buffer, 0, 0);
+                if (err == NO_ERROR)
+                {
+                    m_serverUrlBase = buffer;
```

Comment:
Here you're assigning the `buffer` which is on the stack to `m_serverUrlBase`, `buffer` will get freed as the stack pops after function call, then `m_serverUrlBase` may become dangling pointer

Reply:
Fixed